appfl/algorithm/fedavg.py

```diff
@@ -4,71 +4,97 @@
+                ## computation
+                tmp += self.weights[i] * self.primal_states[i][name]
+
+            global_state[name] = tmp
+
+        """ model update """
+        self.model.load_state_dict(global_state)
 
-        self.model.load_state_dict(update_state)
 
 
 class FedAvgClient(BaseClient):
-    def __init__(
-        self, id, model, optimizer, optimizer_args, dataloader, device, **kwargs
-    ):
-        super(FedAvgClient, self).__init__(
-            id, model, optimizer, optimizer_args, dataloader, device
-        )
+    def __init__(self, id, weight, model, dataloader, device, **kwargs):
+        super(FedAvgClient, self).__init__(id, weight, model, dataloader, device)
         self.loss_fn = CrossEntropyLoss()
         self.__dict__.update(kwargs)
         self.id = id
-
-    # update local model
+ 
     def update(self):
+
         self.model.train()
         self.model.to(self.device)
-        optimizer = self.optimizer(self.model.parameters(), **self.optimizer_args)
+
+        optimizer = eval(self.optim)(self.model.parameters(), **self.optim_args)
+
+        """ Inputs """                
+        ## "global_state" from a server is already stored in 'self.model'
         
+        """ Multiple local update """
         for i in range(self.num_local_epochs):
-            # log.info(f"[Client ID: {self.id: 03}, Local epoch: {i+1: 04}]")
-
             for data, target in self.dataloader:
+
                 data = data.to(self.device)
                 target = target.to(self.device)
+
                 optimizer.zero_grad()
                 output = self.model(data)
-                loss = self.loss_fn(output, target)                
-                loss.backward()                
+                loss = self.loss_fn(output, target)
+                loss.backward()
+
                 optimizer.step()
-        
-        ## Differential Privacy
-        if self.privacy == True:            
-            # Note: Scale_value = Sensitivity_value / self.epsilon            
-            
-            Scale_value = self.scale_value             
-
-            for name, param in self.model.named_parameters():                       
-                mean  = torch.zeros_like(param.data)
-                scale = torch.zeros_like(param.data) + Scale_value
-                m = torch.distributions.laplace.Laplace( mean, scale )                    
-                param.data += m.sample()         
 
-        # self.model.to("cpu")
+        """ Differential Privacy """
+        if self.privacy == True:
+            # Note: Scale_value = Sensitivity_value / self.epsilon
+
+            Scale_value = self.scale_value
+
+            for name, param in self.model.named_parameters():
+                mean = torch.zeros_like(param.data)
+                scale = torch.zeros_like(param.data) + Scale_value
+                m = torch.distributions.laplace.Laplace(mean, scale)
+                param.data += m.sample()
```

Comment:
This looks like output perturbation, correct? Please leave some descriptions about DP.

Reply:
Fixed. I added the description on "BaseClient"